File: algorithms/medium/minimum_deletions_to_make_character_frequencies_unique.ex
defmodule Solution do
  @spec min_deletions(s :: String.t()) :: integer
  def min_deletions(s) do
    s
    |> String.to_charlist()
    |> Enum.reduce(%{}, fn c, acc ->
      Map.update(acc, c, 1, &(&1 + 1))
    end)
    |> Map.values()
    |> Enum.sort(&>=/2)
    |> then(fn vs ->
      Enum.reduce_while(vs, {100_000, 0}, fn v, {pre, ct} ->
        cond do
          v == 0 ->
            {:halt, {pre, ct}}

          v >= pre ->
            pre = max(0, pre - 1)
            {:cont, {pre, ct + v - pre}}

          true ->
            {:cont, {v, ct}}
        end
      end)
      |> elem(1)
    end)
  end
end

File: algorithms/medium/minimum_deletions_to_make_character_frequencies_unique.rs
impl Solution {
    pub fn min_deletions(s: String) -> i32 {
        let mut map = vec![0; 26];
        
        for b in s.as_bytes() {
            map[(b - b'a') as usize] += 1;
        }

        map.sort_unstable();
        
        let mut ans = 0;
        
        for i in (0..25).rev() {
            if map[i] >= map[i+1] {
                let t = map[i];
                map[i] = 0.max(map[i+1] - 1);
                ans += t - map[i];
            }
        }
        
        ans
    }
}
